Do not try heurisch if integral contains abs or sign

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
According to @jksuom, see https://github.com/sympy/sympy/issues/16217#issuecomment-471257311, `heurisch` cannot handle integrals with Abs or sign in them, so I disabled that step in case the expression has Abs or sign. Still possible to override by using `heurisch=True`.

#### Other comments
`test_V14` went from a few minutes to a few seconds with the same output on my computer. (The integration part, simplifying takes a few seconds as well.)

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* integrals
    * Integrals with abs or sign will not use the `heurisch` step by default.
<!-- END RELEASE NOTES -->

oscargus committed:
Do not try heurisch if integral contains abs or sign

asmeurer commented:
I thought heurisch could handle abs if it knows that the derivative is abs(x)/x.

oscargus commented:
I have no idea. But the tests pass (locally) and things are much faster.

asmeurer commented:
I tried

```patch
diff --git a/sympy/integrals/heurisch.py b/sympy/integrals/heurisch.py
index 78b6fdbbdd..b1621ec75a 100644
--- a/sympy/integrals/heurisch.py
+++ b/sympy/integrals/heurisch.py
@@ -11,7 +11,7 @@
 from sympy.core.relational import Eq, Ne
 from sympy.core.singleton import S

-from sympy.functions import exp, sin, cos, tan, cot, asin, atan
+from sympy.functions import exp, sin, cos, tan, cot, asin, atan, Abs
 from sympy.functions import log, sinh, cosh, tanh, coth, asinh, acosh
 from sympy.functions import sqrt, erf, erfi, li, Ei
 from sympy.functions import besselj, bessely, besseli, besselk
@@ -258,6 +258,9 @@ def get_diff(self, f):
         elif (not hasattr(f, 'func') or
             not _bessel_table.has(f.func)):
             cache[f] = cancel(f.diff(self.x))
+        elif isinstance(f, Abs):
+            z = f.args[0]
+            cache[f] = Abs(z)/z*self.get_diff(z)
         else:
             n, z = f.args
             d0, d1 = _bessel_table.diffs(f.func, n, z)
```

For me it makes `integrate(x*abs(9-x**2), x)` take 49s, down from 1 min 21s in master, but it doesn't actually give an answer. 

I'm not surprised tests pass. Abs hasn't given that derivative in a long time and I'm pretty sure it has no hope of working unless it does.

asmeurer commented on sympy/integrals/integrals.py:
```diff
@@ -1026,16 +1026,19 @@ def _eval_integral(self, f, x, meijerg=None, risch=None, manual=None,
-                        # implementation of heuristic Risch integration
-                        # algorithm.
-                        h = None
+                    if heurisch is True or not (g.has(Abs) or g.has(sign)):
```
It would be better to have this condition in heurisch itself.

oscargus replied:
Problem is that you cannot force then. At least not without introducing a new flag. But I see the point.

asmeurer replied:
heurisch should be responsible for knowing what sorts of things it can and cannot handle. That is how every other integrator works (except for the polynomial and rational function ones it looks like, but even those could be made to work this way). There could be some internal heurisch function that bypasses this if someone *really* wants to see if something works anyway.

oscargus commented:
Enabling it to work would of course be even better. 

At the moment, it just seems like there are lots of failing integral tests that takes quite a bit of time. (Not to mention when it is actually used...)

oscargus commented:
With this PR:
```
In [109]: %timeit integrate(x*abs(9-x**2), x)
537 ms ± 4.35 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

asmeurer commented:
The thing that I thought would work to enable it didn't work, so I wouldn't worry about that for now.

asmeurer commented:
Actually I didn't realize our abs support in integrate was quite this bad. Even `integrate(abs(x))` and `integrate(sign(x))` do not currently work. Clearly we need better support, but I'm not sure which algorithm(s) that should come from.

jksuom commented:
I think that heurisch can work with complex analytic (holomorphic or meromorphic) functions only. So there should be no abs, re, im. Piecewise analytic functions like Heaviside and sign could probably be handled by the wrapper if it can rewrite them as Piecewise and serve the pieces to heurisch.

smichr commented:
> if it can rewrite them as Piecewise and serve the pieces to heurisch

I take this to mean that this PR is ok to go and that getting functions with Abs to work via heurisch will require more prep-work on the expression being integrated. So this can be committed?

jksuom commented:
I think that all non-analytic functions should be tested: `not g.has(Abs, re, im)`. Currently, also piecewise analytic functions like `sign` and `Heaviside` should be excluded though it might be possible to let `heurisch_wrapper` handle them piece by piece.

oscargus commented:
I think that the test should be moved inside heurisch as @asmeurer suggested. Makes sense, but just haven't figured out what the best way to force it to ignore the fact that it cannot handle these functions...

oscargus committed:
Bypassed inside heurisch instead and added more functions

oscargus commented:
I've modified it now and also added re, im, Heaviside, and DiracDelta to functions that it cannot handle.

All (non-slow) integral tests pass if nothing else.

oscargus commented:
Btw, `integrate(abs(x))` and `integrate(sign(x))` works if `x` is (extended) real.

smichr commented:
> works if `x` is (extended) real.

Perhaps because when x is real it can be rewritten in terms of Piecewise.

If rewrite would return a Piecewise indicating the real result and the unevaluated results it would work for vanilla x:
```python
>>> var('x')
x
>>> Piecewise(
...         (Piecewise((x, x >= 0), (-x, True)), Eq(im(x),0)),  # when real
...         (abs(x),True))  # otherwise
Piecewise((Piecewise((x, x >= 0), (-x, True)), Eq(im(x), 0)), (Abs(x), True))
>>> integrate(_, x)
Piecewise((-x**2/2, x <= 0), (x**2/2, True))
```

oscargus commented:
Yes, there seems to be many things that can be changed when it comes to `Abs`.

However, I am not sure about integrals of piecewises in general, at least not using piecewise_fold and more than one integration variable. But that is another issue...

oscargus commented:
I tried using the rewrite as suggested by @smichr above in several functions and it seems to work. However, it also seems like there are some assumptions on real-valued integral variables:
```
In [32]: re(x).rewrite(Piecewise)
Out[32]: 
⎧  x    for im(x) = 0
⎨                    
⎩re(x)    otherwise  

In [33]: integrate(re(x).rewrite(Piecewise), x)
Out[33]: 
 2
x 
──
2 
```

I do not know if it expected to by like this?

On the other hand:
```
In [22]: integrate(x*abs(9-x**2).rewrite(Piecewise), x)
Out[22]: 
⎧    4      2                
⎪   x    9⋅x                 
⎪   ── - ────      for x ≤ -3
⎪   4     2                  
⎪                            
⎪   4      2                 
⎪  x    9⋅x    81            
⎨- ── + ──── - ──  for x ≤ 3 
⎪  4     2     2             
⎪                            
⎪    4      2                
⎪   x    9⋅x                 
⎪   ── - ────      otherwise 
⎪   4     2                  
⎩                            
```
which earlier didn't evaluate.

So the question is if one should instead rewrite functions including at least abs as Piecewise? Or definite integrals only? DiracDelta doesn't seem to work either way, but probably Heaviside and sign could be rewritten. And maybe re, im?

Another indication that it seems like the integration variable is assumed real-valued at some stage:

```
In [34]: integrate(im(x).rewrite(Piecewise), x)
Out[34]: 
⌠         
⎮ im(x) dx
⌡         

In [35]: integrate(im(I*x).rewrite(Piecewise), x)
Out[35]: 
 2
x 
──
2 
```

jksuom commented:
I don't think that `always_try_heurisch` can be used for anything else than possibly some experimentation. How about writing `_try_heurisch` to hint that it is not for general use?
 
`DiracDelta` is an obvious addition to the exclude list. `arg`, `floor` and `ceiling` should also be there.

oscargus commented:
> I don't think that `always_try_heurisch` can be used for anything else than possibly some experimentation. How about writing `_try_heurisch` to hint that it is not for general use?

Me neither. Makes sense!

> `DiracDelta` is an obvious addition to the exclude list. `arg`, `floor` and `ceiling` should also be there.

I'll add them.

Thanks!

oscargus committed:
Changed name of argument and more functions that are ignored

oscargus commented:
I've finally got around to update this.

jksuom commented:
This looks good. Probably many tests will take less time.

asmeurer commented:
rewriting would work at the integrate level but not at the heurisch level. Heurisch doesn't know how to deal with Piecewise either, but there are Piecewise algorithms in integrate that split up Piecewise and recursively integrate the separate parts.
